blitbuffer.c

```diff
@@ -456,7 +456,7 @@ void BB_blit_to_BB8(BlitBuffer *src, BlitBuffer *dst,
                 // (i.e., setPixel, no rota, no invert).
                 // The cbb codepath ensures setPixel & no invert, so we only check for rotation.
                 if (sbb_rotation == 0 && dbb_rotation == 0) {
-                    if (offs_x == 0 && dest_x == 0 && w == src->phys_w && w == dst->phys_w) {
+                    if (offs_x == 0 && dest_x == 0 && w >= src->w && w >= dst->w && src->pitch == dst->pitch) {
```

Comment:
In that one, w *has* to match the stride in pixels (of both src & dst). It's a full-screen blit of w * h, so if w doesn't match the stride, it'll either copy slightly too much or slightly not enough data.

Which, granted, might be nearly unnoticeable in the latter and possibly harmless in the former, but, still ^^.

Reply:
Okay, >= means that it could only copy too *much* data, granted.

That may be harmless on Y8, but it won't be on RGB32, where we usually do *NOT* have any scratch space past the screen.

Reply:
w is always <= pitch. If there's no scratch space, then w==pitch. Additional w<pitch check could be done, but I'm simply assuming the given rect dimensions are always within pitch from the caller (is that a valid assumption though?)

Reply:
I vaguely recall PDFs *maybe* doing some super-wonky things like sometimes passing the document's width, which may be very very large indeed, but I *may* be wrong.

Reply:
I'm vaguely wondering if the simplest solution is not simply to drop the w checks entirely, and simply use pitch*h as the copy size?

Reply:
Wait, no, that's dumb, we do need to ensure that we actually requested a blit of at least a scanline's worth of pixels for that to work.

Reply:
Well, technically, a scanline's worth of pixels *exactly*, which was my original point ^^.

Reply:
c.f., https://github.com/koreader/koreader-base/commit/6e7cd40b9425926ef22d8bcb6527396b910516f7#commitcomment-42580213; I think checkBounds takes care of w <= pitch for us ;).

Reply:
The idea here is "be as permissive as possible without breaking stuff" but that indeed probably isn't wise, as even the old code managed to get it wrong. Perhaps just forgo this altogether except the trivial safe case (x=0,w=dst->w,pitch==pitch) as the *practical* speed up of block copy compared to copy-per-scanline are most likely academic (sub 0.1%) when we're copying near full scanline length anyway..

Reply:
Agreed ;).

And, yeah, I did test that at the time, and the difference between a screen copy vs. a height * scanlines copy are indeed minimal ;).

But I like it nonetheless ;p.